core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,152 @@
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * File cleanup strategy for snapshot expiration which determines, via an in-memory reference set,
+ * metadata and data files that are not reachable given the previous and current table states.
+ */
+class ReachableFileCleanup extends FileCleanupStrategy {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReachableFileCleanup.class);
+
+  ReachableFileCleanup(
+      FileIO fileIO,
+      ExecutorService deleteExecutorService,
+      ExecutorService planExecutorService,
+      Consumer<String> deleteFunc) {
+    super(fileIO, deleteExecutorService, planExecutorService, deleteFunc);
+  }
+
+  @Override
+  public void cleanFiles(TableMetadata beforeExpiration, TableMetadata afterExpiration) {
+    Set<String> manifestListsToDelete = Sets.newHashSet();
+
+    Set<Snapshot> snapshotsBeforeExpiration = Sets.newHashSet(beforeExpiration.snapshots());
+    Set<Snapshot> snapshotsAfterExpiration = Sets.newHashSet(afterExpiration.snapshots());
+    Set<Snapshot> expiredSnapshots = Sets.newHashSet();
+    for (Snapshot snapshot : snapshotsBeforeExpiration) {
+      if (!snapshotsAfterExpiration.contains(snapshot)) {
+        expiredSnapshots.add(snapshot);
+        if (snapshot.manifestListLocation() != null) {
+          manifestListsToDelete.add(snapshot.manifestListLocation());
+        }
+      }
+    }
+
+    Set<ManifestFile> candidateManifestFilesForDeletion = readManifests(expiredSnapshots);
+    Set<ManifestFile> manifestFilesAfterExpiration = readManifests(snapshotsAfterExpiration);
+
+    Set<ManifestFile> manifestsToDelete = Sets.newHashSet(candidateManifestFilesForDeletion);
+    for (ManifestFile candidateManifestFile : candidateManifestFilesForDeletion) {
+      if (manifestFilesAfterExpiration.contains(candidateManifestFile)) {
+        manifestsToDelete.remove(candidateManifestFile);
+      }
+    }
+
+    Set<String> dataFilesToDelete =
+        findFilesToDelete(manifestsToDelete, manifestFilesAfterExpiration);
+    deleteFiles(dataFilesToDelete, "Data");
+
+    Set<String> manifestPathsToDelete =
+        manifestsToDelete.stream().map(ManifestFile::path).collect(Collectors.toSet());
+
+    deleteFiles(manifestPathsToDelete, "manifest");
+    deleteFiles(manifestListsToDelete, "manifest list");
+  }
+
+  private Set<ManifestFile> readManifests(Set<Snapshot> snapshots) {
+    Set<ManifestFile> manifestFiles = Sets.newHashSet();
+    for (Snapshot snapshot : snapshots) {
+      try (CloseableIterable<ManifestFile> manifestFilesForSnapshot = readManifestFiles(snapshot)) {
```

Comment:
We may want to do this in parallel using `ParallelIterable`. The results are copied anyway so there isn't much of a cost to it.

Reply:
Oh missed this when addressing the comments, is this blocking or could I address in a follow on PR along with https://github.com/apache/iceberg/pull/5669/files#r990877250 ?

Reply:
We can do it in a follow up.

Reply:
In fact, we may want to do something like what you do in `findFilesToDelete`, where rather than reading a set of current manifest files, you just remove manifests from the candidate set. You could also use the same logic to short circuit if there are no more candidate manifests.